### PR TITLE
doc: an ADR about experimental features

### LINKED ISSUE #5862: Proposal: how to add experimental features to existing classes
## Background

From time-to-time we need to add experimental features to existing libraries. These may be:

- improvements or additions that do not have a stable API, normally we have a period with new libraries where we let the API stabilize, but we have no such luxury when adding APIs to a GA library.
- features of a service that are released as "beta" or "EAP", or otherwise not GA, we may want to give folks access to that feature, but again the API might change, or the underlying feature may be completely removed.

Making changes to GA libraries *implies*, unless we take additional measures, that the changes are GA.

The `[[deprecated]]` attribute is not available in C++11, but even if we used the compiler-specific attributes, or `ABSL_DEPRECATED()` the generated warning messages are misleading (they include the literal word `deprecated` in the warnings).

Sometimes it is possible to add new features in a separate namespace, such as `google::cloud::experimental`, or `google::cloud::${service}_experimental`, but that is often cumbersome for features added to a core class like `google::cloud::${service}::Client` or `google::cloud::${service}::Connection`.

## Proposal

Experimental APIs will take a dummy parameter of type `google::cloud::Experimental`.  So all uses of an experimental API will look as:

```cc
namespace spanner = ::google::cloud::spanner;  // Using spanner for illustration purposes
void F(spanner::Client client) {
  auto callback = ....;
  client.AsyncExecuteQuery(
    spanner::SqlStatement(...), callback, google::cloud::Experimental() /*, spanner::QueryOptions*/);
}
```

Once the API becomes stable we can add an overload that does not require the `Experimental` parameter, and mark the experimental option as deprecated.  After a year (or longer), we can remove the overload that requires the `Experimental` parameter.

The `Experimental` class should be fairly simple:

```cc
struct Experimental {
  explicit Experimental() = default;
};
```

## Alternatives considered

We could name the functions "Experimental*", as in `ExperimentalAsyncExecuteQuery()`. Labeling via parameters seems more expressive. We could add other labels, like `internal::ImplementationOnly()`, or `internal::TestOnly()`, and so on.

### PR DESCRIPTION
**Pull request by coryan**
Fixes #5862

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8898)

### CODE REVIEW

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+**Title**: customers must type `[Ee]xperimental` to access experimental features.
```
Comment (dbolduc) [inline doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md:1]: what about a `GOOGLE_CLOUD_CPP_EXPERIMENTAL_FOO=...` environment variable?

Comment (coryan) [reply doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md]: Fixed in the "decision" section

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+- Features of a service that are released as "beta" or "EAP", or otherwise not
+  GA, we may want to give folks access to that feature, but again the API might
+  change, or the underlying feature may be completely removed.
+
+Making changes to GA libraries *implies*, unless we take additional measures,
+that the changes are GA.
+
+The `[[deprecated]]` attribute is not available in C++11, but even if we use
+existing workarounds (e.g. `ABSL_DEPRECATED()`), the generated warning messages
+are misleading: they include the literal word `deprecated` in the warnings.
```
[the remark this answers is not in the document]
Comment (coryan) [reply doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md]: Done.

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+**Decision**: Experimental features for in existing GA libraries **must** 
+require the application developer to type `experimental` or `Experimental` to
+use them.
+
+Examples of acceptable approaches include:
+
+- Adding any new symbols to the `google::cloud::${service}_experimental`
+  namespace.
+
```
[the remark this answers is not in the document]
Comment (coryan) [reply doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md]: Fixed.

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+Sometimes it is possible to add new features in a separate namespace, such as
+`google::cloud::experimental`, or `google::cloud::${service}_experimental`.
+That is often cumbersome for features added to a core class like
+`google::cloud::${service}::Client` or `google::cloud::${service}::Connection`.
+
+We could use a separate branch for experimental features, but then customers
+would need to manually diff the branch to understand what is experimental and
+what is stable. And we would need to support twice as much code.
+
+**Decision**: Experimental features for in existing GA libraries **must**
```
[the remark this answers is not in the document]
Comment (coryan) [reply doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md]: Fixed

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
+**Title**: customers must type `[Ee]xperimental` to access experimental features.
+
+**Status**: accepted
+
+**Context**: From time-to-time we need to add experimental features to existing
+libraries. These may be:
+
+- Improvements or additions that do not have a stable API, normally we have a
```
[the remark this answers is not in the document]
Comment (coryan) [reply doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md]: Used a period instead.

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  period with new libraries where we let the API stabilize, but we have no such
+  luxury when adding APIs to a GA library.
+- Features of a service that are released as "beta" or "EAP", or otherwise not
+  GA, we may want to give folks access to that feature, but again the API might
```
[the remark this answers is not in the document]
Comment (coryan) [reply doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md]: Ditto

**Review by dbolduc — APPROVED** [review]
LGTM

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+**Decision**: Experimental features in existing GA libraries **must** require
+the application developer to type `experimental` (or `Experimental`, or whatever
+casing is correct under the applicable naming convention) to use them.
+
+Examples of acceptable approaches include:
+
+- Adding any new symbols to the `google::cloud::${service}_experimental`
+  namespace.
+- Naming the symbol `ExperimentalFoo` or similar.
+- Functions can require a parameter of type `google::cloud::ExperimentalTag`.
```
Comment (devjgm) [inline doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md:40]: Maybe add that this parameter must not be defaulted either. Maybe that's what you meant by "require"? Consider being extra explicit w/ something like "... require a non-defaulted parameter of ..."

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+We could use a separate branch for experimental features, but then customers
+would need to manually diff the branch to understand what is experimental and
+what is stable. And we would need to support twice as much code.
+
+**Decision**: Experimental features in existing GA libraries **must** require
+the application developer to type `experimental` (or `Experimental`, or whatever
+casing is correct under the applicable naming convention) to use them.
+
+Examples of acceptable approaches include:
```
Comment (devjgm) [inline doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md:35]: Do we want to add anything about build-target names also including "experimental" in their names? Or maybe that's out of scope for this doc.

**Review by devjgm — APPROVED** [review]
Two small comments, but overall, this LGTM. Thanks!

**Review by scotthart — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8898)* status: all files reviewed, 8 unresolved discussions (waiting on @coryan and @dbolduc)

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md (hunk 8 above)
Comment (coryan) [reply doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md:35]: Done.

File: doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md (hunk 7 above)
Comment (coryan) [reply doc/adr/2022-05-06-customers-must-type-experimental-to-access-experimental-features.md:40]: Done.

**Review by devbww — APPROVED** [review]

**Pull request merged**